moflask/jwt.py

```diff
@@ -0,0 +1,127 @@
+"""JWT session handling.
+
+This module provides the common code-base needed for consuming auth-app JWTs.
+
+The JWT’s claims provide a user identifier, an organization and the current roles of the user.
+"""
+
+import functools
+from typing import Iterable
+
+import flask
+import flask_jwt_extended
+from flask_jwt_extended.exceptions import NoAuthorizationError
+
+ctx_stack = flask._request_ctx_stack  # pylint: disable=protected-access
+get_current_session = flask_jwt_extended.get_current_user
+
+
+class Session:
+    """Object that represents a user’s session for a specific organization."""
+
+    @classmethod
+    def from_raw_token(cls, token: dict):
+        """Create a session from raw JWT token data."""
+        claims = token["user_claims"]
+        return cls(token["identity"], claims["org"], claims["roles"])
+
+    def __init__(self, identity: str, org: str, roles: Iterable[str] = None):
+        """Create a new session instance."""
+        self.identity = identity
+        self.org = org
+        self.roles = set(roles or [])
+
+    def push_to_context(self):
+        """Set request context variables for the authenticated user.
+
+        Some apps don’t need anything besides the raw session so this is just a stub per default.
+        """
+
+    def has_any_role_of(self, roles: Iterable[str]):
+        """Check if the session has any of the given roles."""
+        return any(role in self.roles for role in roles)
+
+    def to_token_data(self):
+        """Return the session object turned into the token data structure.
+
+        This is the inverse of from_raw_token().
+        """
+        return {
+            "identity": self.identity,
+            "user_claims": {
+                "org": self.org,
+                "roles": list(self.roles),
+            },
+        }
+
+
+class Manager(flask_jwt_extended.JWTManager):
+    """A JWTManager subclass with a bit of custom behaviour."""
+
+    session_cls = Session
+
+
+manager = Manager()
```

Comment:
I’ve reconsidered how this should work entirely … and made it more work like the other parts of flask-jwt-extended. I’m not awfully fond of their way of registering callbacks instead of sub-classes, but consistency seems nice too.

Reply:
Exactly. I’ve just pushed a commit that does this for e2t-api.